### PR TITLE
More Altair fork transition tests

### PR DESCRIPTION
**Pull request by ralexstokes**
Adds the following tests to further exercise various chain conditions leading into the Altair fork boundary:

- [x] finality across the fork
- [x] randomize attestation participation across the fork boundary
- [x] attestation participation but no finality
- [x] finality but only after the fork

~edit: intend to add the second test before wanting to merge this~

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 1]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+    assert current_epoch < fork_epoch
+    assert current_epoch == spec.GENESIS_EPOCH
+
+    # regular state transition until fork:
+    fill_cur_epoch = False
+    fill_prev_epoch = True
+    blocks = []
+    for _ in range(current_epoch, fork_epoch - 1):
+        if current_epoch == spec.GENESIS_EPOCH:
+            fill_cur_epoch = True
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:267]: Isn't this going to fill both epoch 0 and epoch 1 with attestations from epoch 0? Not that bad, but just a funny consequence.

I might instead just do a `next_epoch_via_block` before the loop to get out of genesis epoch and then just do a prev_epoch fill from there with none of this exceptional logic to change and unchange these booleans

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 2]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+    for _ in range(4):
+        _, blocks_in_epoch, state = next_slots_with_attestations(post_spec,
+                                                                 state,
+                                                                 post_spec.SLOTS_PER_EPOCH,
+                                                                 fill_cur_epoch,
+                                                                 fill_prev_epoch)
+        blocks.extend([post_tag(block) for block in blocks_in_epoch])
+
+    assert state.slot % post_spec.SLOTS_PER_EPOCH == 0
+    assert post_spec.compute_epoch_at_slot(state.slot) == fork_epoch + 4
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:303]: prefer `get_current_epoch`

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 3]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+    current_epoch = spec.get_current_epoch(state)
+    assert current_epoch < fork_epoch
+    assert current_epoch == spec.GENESIS_EPOCH
+
+    # regular state transition until fork:
+    fill_cur_epoch = False
+    fill_prev_epoch = True
+    blocks = []
+    for _ in range(current_epoch, fork_epoch - 1):
+        if current_epoch == spec.GENESIS_EPOCH:
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:350]: Ditto -- consider not doing the boolean switches and just transitioning past genesis epoch with a single block

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 4]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+
+@fork_transition_test(PHASE0, ALTAIR, fork_epoch=3)
+def test_transition_with_random_participation(state, fork_epoch, spec, post_spec, pre_tag, post_tag):
+    """
+    Transition from the initial ``state`` to the epoch after the ``fork_epoch``,
+    including attestations so as to produce finality through the fork boundary.
+    """
+    rng = random.Random(1337)
+
+    def _drop_random_quarter(_slot, _index, indices):
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:331]: Is this the same exact test logic as above but with a different (> 2/3) participation function?

If so, write a `run_transition_test_with_attestations(spec, state, fork_epoch, participation_function, expect_finality=True)` that wraps this logic up to be reusable. Also you can put `expect_finality=False` if you want to do a participation function with less than 2/3 as well

File: tests/core/pyspec/eth2spec/test/helpers/attestations.py [hunk 5]
```diff
@@ -219,35 +219,56 @@ def add_attestations_to_state(spec, state, attestations, slot):
...
+    participation_fn: (slot, committee_index, committee_indices_set) -> participants_indices_set
+    """
     post_state = state.copy()
     signed_blocks = []
-    for _ in range(spec.SLOTS_PER_EPOCH):
+    for _ in range(slot_count):
         block = build_empty_block_for_next_slot(spec, post_state)
         if fill_cur_epoch and post_state.slot >= spec.MIN_ATTESTATION_INCLUSION_DELAY:
             slot_to_attest = post_state.slot - spec.MIN_ATTESTATION_INCLUSION_DELAY + 1
             committees_per_slot = spec.get_committee_count_per_slot(state, spec.compute_epoch_at_slot(slot_to_attest))
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/helpers/attestations.py:237]: Seems like we may be able to make the inner `if` for prev and cur an internal helper function, passing in `slot_to_attest` and return an attestation`. Might help reduce the code duplication

**Review by djrtwo — APPROVED** [review]
nice!

suggest adding a test with 50% attestations taht goes through the transition and doesn't finalize

also suggest adding a test without attestations through the transition but then finalizes in a later epoch

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 6]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
     yield "blocks", blocks
     yield "post", state
+
+
+@fork_transition_test(PHASE0, ALTAIR, fork_epoch=3)
+def test_transition_with_finality(state, fork_epoch, spec, post_spec, pre_tag, post_tag):
+    """
+    Transition from the initial ``state`` to the epoch after the ``fork_epoch``,
+    including attestations so as to produce finality through the fork boundary.
+    """
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:254]: 👍 on having docstring for each test case (which we probably should write more...).

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 7]
```diff
@@ -242,3 +244,164 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+    assert len(blocks) == (fork_epoch + 4) * post_spec.SLOTS_PER_EPOCH
+    assert len(blocks) == len(set(blocks))
+
+    blocks_without_attestations = [block for block in blocks if len(block.message.body.attestations) == 0]
+    # A boundary condition of ``next_slots_with_attestations`` skips
+    # over the block with ``slot == 1``.
+    # And the fork upgrade helper currently does not construct any attestations.
+    assert len(blocks_without_attestations) == 2
+    slots_without_attestations = [b.message.slot for b in blocks_without_attestations]
+    assert set(slots_without_attestations) == set([1, fork_epoch * spec.SLOTS_PER_EPOCH])
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:404]: This section looks duplicate of a section of `test_transition_with_finality` except for `participation_fn=_drop_random_quarter`. What do you think about refactoring it by adding a common helper?

Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:404]: yeah sounds great!

**Comment by ralexstokes** [comment]
believe i've addressed all the above feedback and added some more test cases -- thanks for the review!

File: tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py [hunk 8]
```diff
@@ -234,11 +236,201 @@ def test_transition_only_blocks_post_fork(state, fork_epoch, spec, post_spec, pr
...
+        participation_fn=_drop_random_quarter
+    )
+
+
+@fork_transition_test(PHASE0, ALTAIR, fork_epoch=3)
+def test_transition_with_random_half_participation(state, fork_epoch, spec, post_spec, pre_tag, post_tag):
+    rng = random.Random(2020)
+
+    def _drop_random_half(_slot, _index, indices):
+        # still finalize, but drop some attestations
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/altair/transition/test_transition.py:374]:
```suggestion
        # drop enough attestations to not finalize
```

File: tests/core/pyspec/eth2spec/test/helpers/attestations.py [hunk 9]
```diff
@@ -219,43 +219,81 @@ def add_attestations_to_state(spec, state, attestations, slot):
...
+                                 state,
+                                 fill_cur_epoch,
+                                 fill_prev_epoch):
+    assert state.slot % spec.SLOTS_PER_EPOCH == 0
+
+    return next_slots_with_attestations(spec,
+                                        state,
+                                        spec.SLOTS_PER_EPOCH,
+                                        fill_cur_epoch,
+                                        fill_prev_epoch)
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/helpers/attestations.py:294]: nitpicking
```suggestion
    return next_slots_with_attestations(
        spec,
        state,
        spec.SLOTS_PER_EPOCH,
        fill_cur_epoch,
        fill_prev_epoch,
    )
```

**Review by hwwhww — APPROVED** [review]
LGTM 👍 
Just one minor nitpicking.

**Review by djrtwo — APPROVED** [review]

**Pull request merged**